plugins/gstreamer/src/GStreamerVideoSource.cpp

```diff
@@ -0,0 +1,299 @@
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <charconv>
+#include <cstdlib>
+
+#include <gst/gst.h>
+#include <gst/app/gstappsink.h>
+
+#include "GStreamerVideoSource.h"
+
+namespace neurala
+{
+namespace
+{
+class Sample
+{
+	GstSample* sample;
+	GstBuffer* buffer;
+	GstMapInfo map;
+
+public:
+	explicit Sample(GstSample* sample) noexcept
+	 : sample(sample), buffer(gst_sample_get_buffer(sample))
+	{
+		gst_buffer_map(buffer, &map, GST_MAP_READ);
+	}
+
+	~Sample() noexcept
+	{
+		gst_buffer_unmap(buffer, &map);
+		gst_sample_unref(sample);
+	}
+
+	dto::ImageView imageView(std::size_t width, std::size_t height) const noexcept
+	{
+		return dto::ImageView(dto::ImageMetadata("uint8", width, height, "RGB", "interleaved", "topLeft"), map.data);
+	}
+
+	Sample(const Sample&) = delete;
+	Sample(Sample&&) = delete;
+
+	Sample& operator=(const Sample&) = delete;
+	Sample& operator=(Sample&&) = delete;
+};
+}
+
+struct GStreamerVideoSource::Implementation
+{
+	GstElement* pipeline;
+	GstElement* userPipeline;
+	GstElement* sink;
+
+	std::unique_ptr<Sample> sample;
+};
+
+void*
+GStreamerVideoSource::create(PluginArguments&, PluginErrorCallback&)
+{
+	if(!gst_is_initialized())
+	{
+		gst_init(0, NULL);
+	}
+	return new GStreamerVideoSource("PLACEHOLDER");
+}
+
+void
+GStreamerVideoSource::destroy(void* p)
+{
+	delete static_cast<GStreamerVideoSource*>(p);
+}
+
+GStreamerVideoSource::GStreamerVideoSource(const char* name)
+ : m_implementation(std::make_unique<Implementation>()),
+   m_bufferReady(false),
+   m_frameReady(false),
+   m_endOfStream(false)
+{
+	m_implementation->pipeline = gst_pipeline_new(name);
+
+	{
+		const auto parser = gst_parse_context_new();
+		const auto userPipeline = getenv("NEURALA_GSTREAMER_PIPELINE");
+
+		if (!userPipeline)
+		{
+			m_lastError = B4BError::invalidParameter();
+			return;
+		}
+
+		m_implementation->userPipeline = gst_parse_launch(userPipeline, nullptr);
+
+		gst_parse_context_free(parser);
+	}
+
+	{
+		long width;
+		long height;
+
+		const auto widthOption = getenv("NEURALA_GSTREAMER_WIDTH");
+		const auto heightOption = getenv("NEURALA_GSTREAMER_HEIGHT");
+
+		if (!widthOption || !heightOption)
+		{
+			m_lastError = B4BError::invalidParameter();
+			return;
+		}
+
+		static constexpr auto ok = std::errc();
+
+		if (std::from_chars(widthOption, widthOption + strlen(widthOption), width).ec != ok
+		    || std::from_chars(heightOption, heightOption + strlen(heightOption), height).ec != ok)
+		{
+			m_lastError = B4BError::invalidParameter();
+			return;
+		}
+
+		m_width = width;
+		m_height = height;
+	}
+
+	m_implementation->sink = gst_element_factory_make("appsink", "sink");
```

Comment:
Why is that created then?

Reply:
Because I didn't push the above change, yet :stuck_out_tongue: